tests/tests/integration_tests.rs

```diff
@@ -414,6 +419,71 @@ async fn ws_server_cancels_stream_after_reset_conn() {
 	assert_eq!(Some(()), rx.next().await, "subscription stream should be terminated after the client was dropped");
 	assert_eq!(Some(()), rx.next().await, "subscription stream should be terminated after the client was dropped");
 }
+// TODO: (dp) Finish this test: check that dropping one subscriber doesn't halt the stream etc
+#[tokio::test]
+async fn ws_server_subscribe_with_stream() {
+	tracing_subscriber::FmtSubscriber::builder()
+		.with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+		.try_init()
+		.expect("setting default subscriber failed");
+
+	use futures::{channel::mpsc, SinkExt, StreamExt};
+	use jsonrpsee::{ws_server::WsServerBuilder, RpcModule};
+
+	let server = WsServerBuilder::default().build("127.0.0.1:0").await.unwrap();
+	let server_url = format!("ws://{}", server.local_addr().unwrap());
+
+	let (tx, mut rx) = mpsc::channel(1);
+	let mut module = RpcModule::new(tx);
+
+	module
+		.register_subscription("subscribe_10_ints", "n", "unsubscribe_10_ints", |_, sink, mut tx| {
+			use futures::task::Poll;
+			let mut int_counter = 1usize;
+			let stream = futures::stream::poll_fn(move |_| -> Poll<Option<usize>> {
+				if int_counter == 10 {
+					return Poll::Ready(None);
+				}
+				std::thread::sleep(Duration::from_millis(100));
+				let out = Poll::Ready(Some(int_counter));
+				int_counter += 1;
+				out
+			});
+
+			tokio::spawn(async move {
+				sink.pipe_from_stream(stream).await.unwrap();
+				let send_back = Arc::make_mut(&mut tx);
+				send_back.feed(()).await.unwrap();
+			});
+			Ok(())
+		})
```

Comment:
@niklasad1 I think something's wrong here but I'm not sure if it's my test code (likely) or our code. What I see when I run `RUST_LOG=jsonrpsee_core=trace cargo t ws_server_subscribe_with_stream` is this:

```
running 1 test
2022-02-23T12:20:55.022002Z TRACE jsonrpsee_core::client::async_client: [frontend]: subscribe: "subscribe_10_ints", unsubscribe: "unsubscribe_10_ints"
2022-02-23T12:20:55.023045Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Entering loop
2022-02-23T12:20:55.128179Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:55.233057Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:55.333694Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:55.439302Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:55.542570Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:55.648021Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:55.752710Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:55.857984Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:55.958600Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:55.958987Z DEBUG jsonrpsee_core::server::rpc_module: Closing subscription: Num(7768645738722848) reason: Some(SubscriptionClosed { reason: Server("No close reason provided") })
2022-02-23T12:20:55.961238Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv method_call Response { jsonrpc: TwoPointZero, result: Number(7768645738722848), id: Number(0) }
2022-02-23T12:20:55.961628Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(7768645738722848), result: Number(1) } }
2022-02-23T12:20:55.961767Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(7768645738722848), result: Number(2) } }
2022-02-23T12:20:55.961878Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(7768645738722848), result: Number(3) } }
2022-02-23T12:20:55.961988Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(7768645738722848), result: Number(4) } }
2022-02-23T12:20:55.962182Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(7768645738722848), result: Number(5) } }
2022-02-23T12:20:55.962431Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(7768645738722848), result: Number(6) } }
2022-02-23T12:20:55.962684Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(7768645738722848), result: Number(7) } }
2022-02-23T12:20:55.962785Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(7768645738722848), result: Number(8) } }
2022-02-23T12:20:55.962875Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(7768645738722848), result: Number(9) } }
2022-02-23T12:20:55.962997Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(7768645738722848), result: Object({"reason": Object({"Server": String("No close reason provided")})}) } }
2022-02-23T12:20:55.963208Z TRACE jsonrpsee_core::client::async_client: [frontend]: subscribe: "subscribe_10_ints", unsubscribe: "unsubscribe_10_ints"
2022-02-23T12:20:55.963844Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Entering loop
2022-02-23T12:20:56.068428Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:56.173776Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:56.274950Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:56.380225Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:56.485690Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:56.587462Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:56.689253Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:56.793708Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:56.897540Z TRACE jsonrpsee_core::server::rpc_module: [SubscriptionSink::pipe_from_stream] Left - sending a result back
2022-02-23T12:20:56.897986Z DEBUG jsonrpsee_core::server::rpc_module: Closing subscription: Num(8581788347571076) reason: Some(SubscriptionClosed { reason: Server("No close reason provided") })
2022-02-23T12:20:56.900452Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv method_call Response { jsonrpc: TwoPointZero, result: Number(8581788347571076), id: Number(2) }
2022-02-23T12:20:56.900874Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(8581788347571076), result: Number(1) } }
2022-02-23T12:20:56.900991Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(8581788347571076), result: Number(2) } }
2022-02-23T12:20:56.901083Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(8581788347571076), result: Number(3) } }
2022-02-23T12:20:56.901174Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(8581788347571076), result: Number(4) } }
2022-02-23T12:20:56.901448Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(8581788347571076), result: Number(5) } }
2022-02-23T12:20:56.901541Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(8581788347571076), result: Number(6) } }
2022-02-23T12:20:56.901655Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(8581788347571076), result: Number(7) } }
2022-02-23T12:20:56.901746Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(8581788347571076), result: Number(8) } }
2022-02-23T12:20:56.902038Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(8581788347571076), result: Number(9) } }
2022-02-23T12:20:56.902565Z DEBUG jsonrpsee_core::client::async_client: [backend]: recv subscription Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayload { subscription: Num(8581788347571076), result: Object({"reason": Object({"Server": String("No close reason provided")})}) } }
test ws_server_subscribe_with_stream ... ok
```

Notice how the two subscriptions seem to each trigger a full run-through of the stream and that this happens sequentially and before the subscriptions on the client side are even read. I wonder if the stream items are buffered up somewhere awaiting subscribers? That would be crazy though, so my test stream must be "wrong" (in the sense that it's not a useful example of a real world stream). What would a better example be?

Also a bit concerning is that when I `mem::forget(sub1)` I would expect to trigger [this](https://github.com/paritytech/jsonrpsee/blob/7351dadf12d5645652290e37164fb037027e6870/core/src/server/rpc_module.rs#L905) but I never see it. How can I write a test that drops the server side `Subscription` and triggers the `Notify` stuff?

Reply:
Fixed by now